### PR TITLE
Check if it has been defined before defining _GNU_SOURCE

### LINKED ISSUE #7957: [Bug]: _GNU_SOURCE macro defined twice in the codebase
### Contact Details

_No response_

### Version

5.7.2

### Description

The `_GNU_SOURCE` macro is defined twice under the exact same `#ifdef` logic in the WolfSSL codebase:

- https://github.com/wolfSSL/wolfssl/blob/398f8c90e22fa8a4edaa50705e7b9896a975f5cb/src/ssl.c#L31
- https://github.com/wolfSSL/wolfssl/blob/398f8c90e22fa8a4edaa50705e7b9896a975f5cb/src/bio.c#L30

This means that you will always get compiler warnings. I believe this is harmless, but adds noise to the log that may distract from actual bugs...

### Reproduction steps

1. Build the library (e.g. embedded in the Netatalk build system)
2. Observe the compiler log

### Relevant log output

```shell
[206/318] Compiling C object libatalk/ssl/libssl.a.p/src_ssl.c.o
In file included from ../libatalk/ssl/wolfcrypt/src/misc.c:29,
                 from ../libatalk/ssl/src/ssl.c:45:
./config.h:800: warning: "_GNU_SOURCE" redefined
  800 | #define _GNU_SOURCE 1
      | 
../libatalk/ssl/src/ssl.c:31: note: this is the location of the previous definition
   31 |     #define _GNU_SOURCE
      | 
In file included from ../libatalk/ssl/src/x509.c:24,
                 from ../libatalk/ssl/src/ssl.c:23296:
./config.h:800: warning: "_GNU_SOURCE" redefined
  800 | #define _GNU_SOURCE 1
      | 
In file included from ../libatalk/ssl/src/ssl.c:21882:
../libatalk/ssl/src/bio.c:30: note: this is the location of the previous definition
   30 |     #define _GNU_SOURCE
      |
```

### PR DESCRIPTION
**Pull request by rdmark**
# Description

Do a check for the `_GNU_SOURCE` macro before defining it.

Fixes #7957

# Testing

Tested by building WolfSSL 5.7.2 against Netatalk.

# Checklist

N/A

### CODE REVIEW

**Comment by dgarske** [comment]
Okay to test. Contributor agreement in progress.

**Comment by dgarske** [comment]
Retest this please:

```
dtls.txt_57
SSL_read input error -501, can't match cipher suite
client failed! Suite = DHE-RSA-AES128-SHA256 version = 3
```

**Review by dgarske — APPROVED** [review]
Do not merge until contributor agreement has been approved.

File: src/bio.c
Comment (douzzer) [inline src/bio.c:29]: remove this now that it's frivolous

File: src/ssl.c
Comment (douzzer) [inline src/ssl.c:30]: here too

**Review by douzzer — CHANGES_REQUESTED** [review]
I assume the problem this addresses is when a system header defines `_GNU_SOURCE` to `1`, then this is undone by the code in `bio.c` and `ssl.c`.  Seems like a good fix.

I'd like to see the `#undef _GNU_SOURCE` clauses removed from inside the conditions, now that the outer gates have made them frivolous.

**Comment by douzzer** [comment]
superseded by #7971

**Pull request closed without merging**
